File: getting-started/todo/java/app/src/main/java/com/amplifyframework/samples/gettingstarted/OptionsBarFragment.java
package com.amplifyframework.samples.gettingstarted;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.RadioButton;
import android.widget.RadioGroup;

import androidx.fragment.app.DialogFragment;

import com.amplifyframework.datastore.generated.model.Priority;
import com.amplifyframework.datastore.generated.model.Todo;
import com.amplifyframework.samples.gettingstarted.databinding.OptionsBarBinding;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import org.jetbrains.annotations.NotNull;

public class OptionsBarFragment extends BottomSheetDialogFragment implements View.OnClickListener {

    private static final String IS_NEW_ITEM = "IsNewItem";
    private static final String POSITION = "Position";
    private static final String ITEM_ADAPTER = "ItemAdapter";
    private static final String ITEM_TEXT = "ItemText";
    private static final String PRIORITY = "Priority";
    private View sheetView;
    private TodoItemAdapter itemAdapter;
    private int position;
    private boolean isNewItem;
    private Priority priority;
    private ImageButton saveBtn;
    private RadioGroup priorityRadioGroup;
    private EditText textBox;

    public OptionsBarFragment(TodoItemAdapter itemAdapter, boolean isNewItem, int position, String text, Priority priority) {
        Bundle b = new Bundle();
        b.putBoolean(IS_NEW_ITEM, isNewItem);
        b.putInt(POSITION, position);
        b.putSerializable(ITEM_ADAPTER, itemAdapter);
        b.putString(ITEM_TEXT, text);
        b.putSerializable(PRIORITY, priority);
        this.setArguments(b);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setStyle(DialogFragment.STYLE_NORMAL, R.style.BottomSheetStyle);
    }

    @Override
    public View onCreateView(
            @NotNull LayoutInflater inflater,
            ViewGroup container,
            Bundle savedInstanceState) {
        OptionsBarBinding _binding = OptionsBarBinding.inflate(inflater, container, false);
        sheetView = _binding.getRoot();
        Bundle b = getArguments();
        if (b != null) {
            itemAdapter = (TodoItemAdapter) b.getSerializable(ITEM_ADAPTER);
            position = b.getInt(POSITION);
            isNewItem = b.getBoolean(IS_NEW_ITEM);
            priority = (Priority) b.getSerializable(PRIORITY);
            String text = b.getString(ITEM_TEXT);
            saveBtn = _binding.saveButton;
            ImageButton priorityBtn = _binding.priorityButton;
            ImageButton trashBtn = _binding.trashButton;
            priorityRadioGroup = _binding.radioGroupPriority;
            textBox = _binding.todoTextEntry;

            if (priority == Priority.LOW) {
                priorityRadioGroup.check(R.id.radioButton_low);
            } else if (priority == Priority.NORMAL){
                priorityRadioGroup.check(R.id.radioButton_med);
            } else if (priority == Priority.HIGH) {
                priorityRadioGroup.check(R.id.radioButton_high);
            }
            textBox.setText(text);

            // Listener for text input in textbox
            textBox.addTextChangedListener(new TextWatcher() {
                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    saveBtn.setEnabled(!s.toString().isEmpty());
                }
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
                @Override
                public void afterTextChanged(Editable s) {}
            });
            saveBtn.setEnabled(!isNewItem);
            saveBtn.setOnClickListener(this);
            priorityBtn.setOnClickListener(this);
            trashBtn.setOnClickListener(this);
        }
        return sheetView;
    }

    // Returns the priority set in RadioGroup, default is Priority.LOW
    private Priority getPriority(View view, RadioGroup priorityRadioGroup, Priority priority) {
        int selectedOption = priorityRadioGroup.getCheckedRadioButtonId();
        RadioButton selectedRadioBtn = view.findViewById(selectedOption);

        switch (selectedRadioBtn.getId()) {
            case (R.id.radioButton_low):
                return Priority.LOW;
            case (R.id.radioButton_med):
                return Priority.NORMAL;
            case (R.id.radioButton_high):
                return Priority.HIGH;
            default:
                return priority;
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case (R.id.save_button) :
                saveButtonBehavior();
                break;
            case (R.id.priority_button):
                priorityButtonBehavior();
                break;
            case (R.id.trash_button):
                trashButtonBehavior();
                break;
            default:
                break;
        }
    }

    private void saveButtonBehavior() {
        if (isNewItem) {
            String todoEntry = textBox.getText().toString();
            priority = getPriority(sheetView, priorityRadioGroup, priority);
            Todo item = itemAdapter.createModel(todoEntry, priority);
            itemAdapter.addModel(item, true);
            itemAdapter.notifyItemInserted(itemAdapter.getItemCount() - 1);
        } else {
            Todo item = itemAdapter.getItem(position);
            String todoEntry = textBox.getText().toString();
            priority = getPriority(sheetView, priorityRadioGroup, priority);
            itemAdapter.setModel(position, itemAdapter.updateModel(item, todoEntry, priority, null));
            itemAdapter.notifyItemChanged(position);
        }
        textBox.getText().clear();
    }

    private void priorityButtonBehavior() {
        if (priorityRadioGroup.getVisibility() == View.GONE) {
            priorityRadioGroup.setVisibility(View.VISIBLE);
        } else {
            priorityRadioGroup.setVisibility(View.GONE);
        }

    }

    private void trashButtonBehavior() {
        if (isNewItem) {
            dismiss();
        } else {
            itemAdapter.deleteModel(position);
            textBox.getText().clear();
            isNewItem = true;
            priority = Priority.LOW;
            priorityRadioGroup.check(R.id.radioButton_low);
        }
    }
}

File: getting-started/todo/kotlin/app/src/main/java/com/amplifyframework/samples/gettingstarted/OptionsBarFragment.kt
package com.amplifyframework.samples.gettingstarted

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageButton
import android.widget.RadioButton
import android.widget.RadioGroup
import androidx.fragment.app.DialogFragment
import com.amplifyframework.datastore.generated.model.Priority
import com.amplifyframework.samples.gettingstarted.databinding.OptionsBarBinding
import com.google.android.material.bottomsheet.BottomSheetDialogFragment

class OptionsBarFragment : BottomSheetDialogFragment(), View.OnClickListener {
    private var _binding: OptionsBarBinding? = null
    private val binding get() = _binding!!
    private lateinit var sheetView: View
    private lateinit var itemAdapter: TodoItemAdapter
    private var position: Int = 0
    private var isNewItem: Boolean = false
    private lateinit var priority: Priority
    private lateinit var saveBtn: ImageButton
    private lateinit var priorityRadioGroup: RadioGroup
    private lateinit var textBox: EditText

    companion object {
        const val IS_NEW_ITEM = "IsNewItem"
        const val POSITION = "Position"
        const val ITEM_ADAPTER = "ItemAdapter"
        const val ITEM_TEXT = "ItemText"
        const val PRIORITY = "Priority"
        fun newInstance(
            itemAdapter: TodoItemAdapter,
            isNewItem: Boolean,
            position: Int,
            text: String,
            priority: Priority
        ): OptionsBarFragment {
            val b = Bundle()
            val optionsBar = OptionsBarFragment()
            b.putBoolean(IS_NEW_ITEM, isNewItem)
            b.putInt(POSITION, position)
            b.putSerializable(ITEM_ADAPTER, itemAdapter)
            b.putString(ITEM_TEXT, text)
            b.putSerializable(PRIORITY, priority)
            optionsBar.arguments = b
            return optionsBar
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setStyle(DialogFragment.STYLE_NORMAL, R.style.BottomSheetStyle)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = OptionsBarBinding.inflate(inflater, container, false)
        sheetView = binding.root
        val b: Bundle? = arguments
        itemAdapter = b?.getSerializable(ITEM_ADAPTER) as TodoItemAdapter
        position = b.getInt(POSITION)
        isNewItem = b.getBoolean(IS_NEW_ITEM)
        priority = b.getSerializable(PRIORITY) as Priority
        val text: String? = b.getString(ITEM_TEXT)
        saveBtn = binding.saveButton
        val priorityBtn: ImageButton = binding.priorityButton
        val trashBtn: ImageButton = binding.trashButton
        priorityRadioGroup = binding.radioGroupPriority
        textBox = binding.todoTextEntry

        when (priority) {
            Priority.LOW -> priorityRadioGroup.check(R.id.radioButton_low)
            Priority.NORMAL -> priorityRadioGroup.check(R.id.radioButton_med)
            Priority.HIGH -> priorityRadioGroup.check(R.id.radioButton_high)
        }
        textBox.setText(text)

        // Listener for text input in textbox
        textBox.addTextChangedListener(
            object : TextWatcher {
                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                    saveBtn.isEnabled = s.toString().isNotBlank()
                }

                override fun beforeTextChanged(
                    s: CharSequence,
                    start: Int,
                    count: Int,
                    after: Int
                ) {
                }

                override fun afterTextChanged(s: Editable) {
                }
            }
        )
        saveBtn.isEnabled = !isNewItem
        saveBtn.setOnClickListener(this)
        priorityBtn.setOnClickListener(this)
        trashBtn.setOnClickListener(this)
        return sheetView
    }

    // Returns the priority set in RadioGroup, default is Priority.LOW
    private fun getPriority(
        view: View,
        priorityRadioGroup: RadioGroup,
        priority: Priority
    ): Priority {
        val selectedOption = priorityRadioGroup.checkedRadioButtonId
        val selectedRadioBtn: RadioButton = view.findViewById(selectedOption)
        return when (selectedRadioBtn.id) {
            R.id.radioButton_low -> Priority.LOW
            R.id.radioButton_med -> Priority.NORMAL
            R.id.radioButton_high -> Priority.HIGH
            else -> {
                priority
            }
        }
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.save_button -> saveButtonBehavior()
            R.id.priority_button -> priorityButtonBehavior()
            R.id.trash_button -> trashButtonBehavior()
        }
    }

    // Saves an item when pressed
    private fun saveButtonBehavior() {
        if (isNewItem) {
            val todoEntry = textBox.text.toString()
            priority = getPriority(sheetView, priorityRadioGroup, priority)
            val item = itemAdapter.createModel(todoEntry, priority)
            itemAdapter.addModel(item, true)
            itemAdapter.notifyItemInserted(itemAdapter.itemCount - 1)
        } else {
            val item = itemAdapter.getItem(position)
            val todoEntry = textBox.text.toString()
            priority = getPriority(sheetView, priorityRadioGroup, priority)
            itemAdapter.setModel(position, itemAdapter.updateModel(item, todoEntry, priority, null))
            itemAdapter.notifyItemChanged(position)
        }
        textBox.text.clear()
    }

    // Opens and closes priorityRadioGroup when pressed
    private fun priorityButtonBehavior() {
        if (priorityRadioGroup.visibility == View.GONE) {
            priorityRadioGroup.visibility = View.VISIBLE
        } else {
            priorityRadioGroup.visibility = View.GONE
        }
    }

    // Deletes an item when pressed
    private fun trashButtonBehavior() {
        if (isNewItem) {
            dismiss()
        } else {
            itemAdapter.deleteModel(position)
            textBox.text.clear()
            isNewItem = true
            priority = Priority.LOW
            priorityRadioGroup.check(R.id.radioButton_low)
            dismiss()
        }
    }
}
